Add genrics store function, modify parse annotation path to search an…

Synicix opened the pull request:
-Add genrics store function
-modify parse annotation path to search annotation
- fix delete pod missing handle case and convert it to use generics
- made get_struct_name panic if fail to extract
- fix regex
- remove regex not found

Synicix committed:
Add genrics store function, modify parse annotation path to search annotation, fix delete pod missing handle case, made get_struct_name panic if fail to extract

Synicix commented:
#26

Synicix committed:
Save progress

Synicix committed:
Change LocalFileStore ref to Self

Synicix committed:
Merge branch 'parse_annotation' into pod-job

Synicix committed:
Add generic save

Synicix committed:
Merge branch 'parse_annotation' into pod-job

Synicix committed:
Fix regex error

Synicix committed:
Save progress

Synicix committed:
Merge branch 'parse_annotation' into pod-job

Synicix commented:
@guzman-raphael @eywalker Going to open it up for discussion on either tomorrow or Friday

Synicix commented:
Note #35 will depend on this, since it uses a lot of the generic functions I made

Synicix committed:
Save progress

Synicix committed:
Remove hardcoded test

Synicix committed:
Fix bug with delete item

Synicix committed:
Merge branch 'parse_annotation' into pod-job

Synicix committed:
Remove hardcoded intergration test

Synicix committed:
Update docs for search_annotation

Synicix committed:
add some addititional docs

eywalker commented on src/util.rs:
```diff
@@ -1,15 +1,14 @@
+    type_name::<T>()
         .split("::")
         .collect::<Vec<&str>>()
         .get(2)
```
Actually why are we getting position 2?

Synicix replied:
Okay just checked the code, since I didn't really looked at it before after it got merge into the initial pull request.

So, for example, given a pod, type_name::<Pod>() returns "orcapod::model::Pod", so we would need to extract the last section and ignore the rest. Given that, this should have been implemented with last and not get(2). I will make a change quickly

eywalker reviewed (changes requested):
I agree with the strict name lookup not returning a result. However I'd want to reassess its use and logic.

Synicix committed:
Forgot to save

Synicix committed:
Add partial eq

Synicix committed:
Change hash to lower case to match other sha256

Synicix committed:
Remove uneeded pod type

Synicix committed:
Add markdown to commets, otherwise cargo test try to run them

Synicix committed:
Fix implmentation for get_struct_name, also renamed to match the behavior.

Synicix committed:
Merge branch 'parse_annotation' into pod-job

Synicix committed:
remove model_test

Synicix committed:
Fix incorrect merging

Synicix committed:
remove todo

Synicix committed:
Conform with clippy

Synicix committed:
Add clippy

Synicix committed:
Move annotation to pod folder and add cacheing mech

Synicix committed:
Add make anno path and move annotation to a seperate folder inside the item

Synicix commented:
@eywalker @guzman-raphael 
Summary:
- Change annotation design to save in the item folder
- Add cacheing in form of BTreeMap<name-ver, item_hash> with binary encoding and race condition handling for file system io

TODO:
- Need to add glob and regex to construct BTreeMap in case there is a change done outside the system.
- Error Handling (might leave till later)

I have updated the pull request to support the new design of annotation with a slight modification to make the folder structure cleaner.

What we agree to try yesterday was to save annotation in the item folder, but instead of doing that, I now save it under all annotations under its own folder, under the item folder.

i.e
/pod/annotations/<name>-<ver>.yaml
The reason behind this because I didn't want to gunk up the main folder with a bunch of annotations that user may not care about.

As for caching, I implemented a BTreeMap<name-ver, item_hash> along with storing it in a binary and the associated file locking system to prevent race conditions when more than one instance of the program is updating the localfilestore.

Synicix commented on .vscode/launch.json:
Revert the change

Synicix replied:
Probably need to be updated

Synicix commented on src/store/filestore.rs:
```diff
@@ -113,97 +47,256 @@ impl LocalFileStore {
     }
 
-    fn make_spec_path(&self, class: &str, hash: &str) -> PathBuf {
+    fn make_path<T>(&self, hash: &str, file_name: &str) -> PathBuf {
```
Look into join

Synicix replied:
did join

Synicix commented on src/store/filestore.rs:
```diff
@@ -113,97 +47,256 @@ impl LocalFileStore {
+        path
+    }
+
+    fn make_anno_path<T>(&self, hash: &str, file_name: &str) -> PathBuf {
```
rename to make_annotation_path

Synicix replied:
reanmed

Synicix commented on src/store/filestore.rs:
```diff
@@ -113,97 +47,256 @@ impl LocalFileStore {
+    }
+
+    /// Generic function for loading spec.yaml into memory
+    fn load_item<T: DeserializeOwned>(
```
change to 
load_item<T:asjdfklajsdf>(&self, spec; ResourceSpec)

enum ResourceSpec {
hash(string,
annotation(String, String)
}

Synicix replied:
updated

Synicix commented on src/store/filestore.rs:
```diff
@@ -113,97 +47,256 @@ impl LocalFileStore {
+        from_yaml::<T>(&spec_yaml, &hash, Some(&annotation_yaml))
+    }
+
+    fn delete_item<T>(&self, name: &str, version: &str) -> Result<(), Box<dyn Error>> {
```
Match ResourceSpec above

Synicix commented on src/store/filestore.rs:
```diff
@@ -113,97 +47,256 @@ impl LocalFileStore {
+        from_yaml::<T>(&spec_yaml, &hash, Some(&annotation_yaml))
+    }
+
+    fn delete_item<T>(&self, name: &str, version: &str) -> Result<(), Box<dyn Error>> {
```
Explicit delete the item regardless if there are annotation pointing to it.

Synicix replied:
Also add delete annotation function

Synicix replied:
^ fine with generic

Synicix replied:
Implemented generic

Synicix commented on src/store/mod.rs:
```diff
@@ -1,10 +1,16 @@
 use crate::model::Pod;
 use std::{collections::BTreeMap, error::Error};
 
+pub struct ItemInfo {
```
Be removed

Synicix replied:
removed

Synicix reviewed (commented):
TODO changes

Synicix committed:
Merge remote-tracking branch 'upstream/dev' into parse_annotation

Synicix committed:
Save progress for now, still need to do test

Synicix committed:
Fix clippy recommmends

Synicix committed:
Add test structure for new design

Synicix committed:
Change to generic for delete_annotation

Synicix committed:
Remove item info

Synicix commented:
Ready for review with the new implementation of cache

Synicix committed:
remove bincode

Synicix committed:
Revert change from util

Synicix committed:
remove unused code

guzman-raphael commented on .vscode/launch.json:
```diff
@@ -33,8 +33,6 @@
                     "--no-run",
                     "--package=orcapod",
-                    "--test",
-                    "model",
```
You should modify this on your side as needed but we shouldn't change what is committed here since it is an example. Changes should be only if there is a bug or if there is an improvement to the example.

guzman-raphael commented on Cargo.toml:
```diff
@@ -30,6 +30,7 @@ sha2 = "0.10.8"
 glob = "0.3.1"
 regex = "1.11.0"
 colored = "2.1.0"
+anyhow = "1.0.90"
```
Rather than adding a dependency for this, you could have added an alias with one line in `error.rs`. Then you could import it everywhere.

For example:

```rust
pub type Result<T> = std::result::Result<T, std::error::Error>;
```

Synicix replied:
Wouldn't that still require you to import Result from error now, when it is not really tied to that? In short anyhow is just a generic wrapper that allow for more stack tracing that takes any error type.

guzman-raphael replied:
We shouldn't rely on an external dependency when we can do this ourselves with just one line. Also, this will be changed anyway once #37 is merged.

Synicix replied:
Perhaps I didn't do a good job of explaining it. The anyhow Error is a wrapper around the std::error that has some additional feature regarding backtrace. It doesn't really change the behavior but instead add on to it.

More info can be found here: https://docs.rs/anyhow/latest/anyhow/

guzman-raphael replied:
@Synicix is looking into why the alias isn't working when set in the library.

guzman-raphael commented on Cargo.toml:
```diff
@@ -50,6 +51,7 @@ restriction = "deny"
 nursery = "deny"
 cargo = "deny"
 
+allow-attributes-without-reason = { level = "allow", priority = 127 }
```
We shouldn't add an exception for this error. All clippy attributes added should include a reason/justification to document why it was necessary for that case.

Synicix replied:
I removed it mainly because of test that would require a lot of these exception when it is pretty obvious as to the reason why. Added it back in

guzman-raphael commented on src/util.rs:
```diff
@@ -1,20 +1,16 @@
-    clippy::unwrap_used,
-    reason = "`None` impossible since `type_name` always returns `&str`."
-)]
+#[expect(clippy::unwrap_used)]
```
See comment above. Reason should not have been removed.

guzman-raphael commented on Cargo.toml:
```diff
@@ -72,6 +74,7 @@ missing_asserts_for_indexing = { level = "allow", priority = 127 }     # missing
 missing_docs_in_private_items = { level = "allow", priority = 127 }    # missing docs on private ok
 missing_inline_in_public_items = { level = "allow", priority = 127 }   # let rust compiler determine best inline logic
 missing_trait_methods = { level = "allow", priority = 127 }            # allow in favor of rustc `implement the missing item`
+multiple_crate_versions = { level = "allow", priority = 127 }
```
I've just tried removing this on my side and there was no error. Perhaps this change is now stale so we should remove it. Also, when adding a new one, we should use a comment to give context why this was necessary.

Synicix replied:
It had to do with fs4 and colorized. I remove fs4 for now since we are not doing file operations, but I can include the reason why

guzman-raphael commented on Cargo.toml:
```diff
@@ -93,9 +96,11 @@ std_instead_of_core = { level = "allow", priority = 127 }              # we shou
 string_add = { level = "allow", priority = 127 }                       # simple concat ok
 string_lit_chars_any = { level = "allow", priority = 127 }             # favor readability until a perf case comes up
 todo = { level = "warn", priority = 127 }                              # warn todos
+verbose_file_reads = { level = "allow", priority = 127 }
```
I've just tried removing this on my side and there was no error. Perhaps this change is now stale so we should remove it. Also, when adding a new one, we should use a comment to give context why this was necessary.

Synicix replied:
Yeah this was an issue with fs4, I removed it because it was complaining that
``` rust
let file = fs::OpenOptions::new().read(true).open("foo.txt");
file.exclusive_lock()
```

should be
```
fs::read(<file>)
```

The issue is that, fs4 requires verbose file reads, and we will most likely run into it later when we start to use that. But for not I can remove it

guzman-raphael commented on Cargo.toml:
```diff
@@ -93,9 +96,11 @@ std_instead_of_core = { level = "allow", priority = 127 }              # we shou
+verbose_file_reads = { level = "allow", priority = 127 }
 use_debug = { level = "warn", priority = 127 }                         # debug print
-
+unseparated_literal_suffix = { level = "allow", priority = 127 }
```
I've just tried removing this on my side and there was no error. Perhaps this change is now stale so we should remove it. Also, when adding a new one, we should use a comment to give context why this was necessary.

Synicix replied:
Odd, before than it was complaining about the regex. Guess when I rebuilt the container it might have became stale with the new standards. Removed.

guzman-raphael commented on Cargo.toml:
```diff
@@ -93,9 +96,11 @@ std_instead_of_core = { level = "allow", priority = 127 }              # we shou
-
+unseparated_literal_suffix = { level = "allow", priority = 127 }
 # temporary
+missing_panics_doc = { level = "allow", priority = 127 }
```
I've just tried removing this on my side and there was no error. Perhaps this change is now stale so we should remove it. Also, when adding a new one, we should use a comment to give context why this was necessary.

Synicix replied:
Issue with this was missing_panic_doc was triggering on the helper functions in test that do panic due to unwrap()
```
error: docs for function which may panic missing `# Panics` section
  --> tests/fixture/mod.rs:25:5
   |
25 |     pub fn get_name(&self) -> &str {
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
note: first possible panic found here
  --> tests/fixture/mod.rs:27:32
   |
27 |             Self::Pod(pod) => &pod.annotation.as_ref().unwrap().name,
   |                                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#missing_panics_doc
   = note: `-D clippy::missing-panics-doc` implied by `-D clippy::pedantic`
   = help: to override `-D clippy::pedantic` add `#[allow(clippy::missing_panics_doc)]`

error: docs for function which may panic missing `# Panics` section
  --> tests/fixture/mod.rs:38:5
   |
38 |     pub fn get_version(&self) -> &str {
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
note: first possible panic found here
  --> tests/fixture/mod.rs:40:32
   |
40 |             Self::Pod(pod) => &pod.annotation.as_ref().unwrap().version,
   |                                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#missing_panics_doc

error: docs for function which may panic missing `# Panics` section
  --> tests/fixture/mod.rs:51:5
   |
51 |     pub fn set_name(&mut self, name: &str) {
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
note: first possible panic found here
  --> tests/fixture/mod.rs:53:52
   |
53 |             Self::Pod(pod) => name.clone_into(&mut pod.annotation.as_mut().unwrap().name),
   |                                                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#missing_panics_doc
```

I added the reason for adding it

guzman-raphael commented on src/store/mod.rs:
```diff
@@ -1,11 +1,13 @@
-    fn load_pod(&self, name: &str, version: &str) -> Result<Pod, Box<dyn Error>>;
-    fn list_pod(&self) -> Result<BTreeMap<String, Vec<String>>, Box<dyn Error>>;
-    fn delete_pod(&self, name: &str, version: &str) -> Result<(), Box<dyn Error>>;
+    fn save_pod(&mut self, pod: &Pod) -> Result<()>;
```
Why was it necessary to make `self` a `mut`? Because of our immutability design, we've generally found ways of not needing this. I'm missing some context/justification here.

Synicix replied:
We have to make it mut, due to self-containing the cache structure which needs to be updated in memory. The previous version stored it on the disk hence why we were able to get away with it.

Synicix replied:
I think discussion can be continue on slack

guzman-raphael commented on src/store/mod.rs:
```diff
@@ -1,11 +1,13 @@
-    fn delete_pod(&self, name: &str, version: &str) -> Result<(), Box<dyn Error>>;
+    fn save_pod(&mut self, pod: &Pod) -> Result<()>;
+    fn load_pod(&mut self, name: &str, version: &str) -> Result<Pod>;
+    fn list_pod(&mut self) -> Result<&BTreeMap<String, String>>;
```
Unclear why the output type changed here. Keep in mind the idea was to use this output to display a table to the user (UI or CLI).

A typical programming pattern is to represent tables as as an ordered mapping where keys are the column name and value is an array of row values (all with the same length). This scales well, is easy to modify, and gives greater flexibility to the caller.

I think I remember you saying that the value might have become a concatenated value of 2 fields? If so, not sure I understand why it is better to always require string splitting of each row by the caller if they need to consume them.

Synicix replied:
Currently it just forward back the cache which is comprise of BtreeString<(name-ver), (hash)> for easy look up and table construction.

I would be okay with converting where keys become the column name. What do you think @eywalker

guzman-raphael commented on src/store/mod.rs:
```diff
@@ -1,11 +1,13 @@
+    fn save_pod(&mut self, pod: &Pod) -> Result<()>;
+    fn load_pod(&mut self, name: &str, version: &str) -> Result<Pod>;
+    fn list_pod(&mut self) -> Result<&BTreeMap<String, String>>;
+    fn delete_pod(&mut self, name: &str, version: &str) -> Result<()>;
```
Assuming `delete_annotation` below is the explicit delete of 1 annotation, would this then be an explicit delete of a whole pod i.e. the pod spec directory + all annotation references?

If so, then it might be better to change the input signature to just require a `hash` argument.

Synicix replied:
delete annotation, will only delete annotation and not the underlying object. Going off of what I remember we discussed last week. Delete_pod will wipe the entire item along with all annotations

guzman-raphael commented on src/error.rs:
```diff
@@ -1,5 +1,4 @@
 use colored::Colorize;
-use serde_yaml;
```
We discussed this at some point with @eywalker. It is isn't required by Rust since you can inline import, however, we mentioned following a good practice for ourselves of including everything that is coming externally at the top through `use` statements for clarity.

Unfortunately clippy doesn't have a lint for the ones that are only 1 level deep...

Synicix replied:
I think the reason why I removed it was because it conflicts with Error thus we need to explicitly define it as serde_yaml::Error making the import moot. I can add it back in for the sake of clairty

guzman-raphael commented on src/store/mod.rs:
```diff
@@ -1,11 +1,13 @@
-    fn list_pod(&self) -> Result<BTreeMap<String, Vec<String>>, Box<dyn Error>>;
-    fn delete_pod(&self, name: &str, version: &str) -> Result<(), Box<dyn Error>>;
+    fn save_pod(&mut self, pod: &Pod) -> Result<()>;
+    fn load_pod(&mut self, name: &str, version: &str) -> Result<Pod>;
```
This PR seems to also address making `Annotation` optional. If so, I would expect `load_pod` to handle loading with name/version (with annotation) or using hash (w/o annotation). The corresponding [issue](https://github.com/walkerlab/orcapod/issues/27) has more info on these cases.

My recommendation would actually be to move that feature into a separate PR since it is starting to become too much for one PR. That way it isn't all-or-nothing.

Synicix replied:
A few of the changes deals provides the ability to do that a bit easier in the future, but since we already decided on going with enum to identify it, I can code it in by tonight.

Synicix replied:
Update pull request to have handle this case along with tests.

guzman-raphael reviewed (changes requested):
@Synicix Thanks for the updates. :handshake:

Submitting PR review progress for now but still need to review the following when I get back.

- src/model.rs
- src/filestore.rs
- tests/fixture/mod.rs
- tests/model.rs
- tests/store.rs

Synicix committed:
revert change in config

Synicix committed:
Added reason back in

Synicix committed:
Edit reason

Synicix committed:
Added reason for exception

Synicix committed:
Remove stale verbose file read

Synicix committed:
Added reason

Synicix committed:
Add case to handle load or delete by hash

Synicix committed:
Merge branch 'parse_annotation' into pod-job

Synicix committed:
Remove caching for now

Synicix commented:
Removed cache for now @guzman-raphael @eywalker  Ready for review

Synicix committed:
Merge branch 'parse_annotation' into pod-job

guzman-raphael commented on src/model.rs:
```diff
@@ -22,31 +17,31 @@ pub fn to_yaml<T: Serialize>(instance: &T) -> Result<String, Box<dyn Error>> {
-    let annotation: Mapping = serde_yaml::from_str(&fs::read_to_string(annotation_file)?)?;
-    let spec_yaml = BufReader::new(fs::File::open(spec_file)?)
-        .lines()
-        .skip(1)
```
I see you removed the skip of the first line which had the `class` attribute. Is it not needed anymore? Is `serde_yaml` smart enough to see that in `T` (`Pod`) it isn't defined and therefore it skips it automatically?

Synicix replied:
Yes, if the field is not defined in the class it skips it and doesn't load it in.

guzman-raphael commented on src/model.rs:
```diff
@@ -22,31 +17,31 @@ pub fn to_yaml<T: Serialize>(instance: &T) -> Result<String, Box<dyn Error>> {
 // --- core model structs ---
 
-#[derive(Serialize, Deserialize, Debug)]
+#[derive(Serialize, Deserialize, Debug, PartialEq, Clone)]
```
What was the reason you needed `PartialEq` and `Clone` on several of these? Was it for asserts in tests?

Synicix replied:
Yep that is the main reason

guzman-raphael commented on src/model.rs:
```diff
@@ -71,9 +66,9 @@ impl Pod {
+    ) -> Result<Self> {
         let pod_no_hash = Self {
-            annotation,
+            annotation: Some(annotation),
```
Rather than doing this, I think it would be better to make `annotation` optional in `new`'s arguments as well so you can initialize a `pod` w/o annotation too. Then, it would should be a pass-through like the others.

Synicix replied:
Fair, since we are going to use the none annotation later on anyways. Will change

guzman-raphael commented on src/store/filestore.rs:
```diff
@@ -1,213 +1,216 @@
+            let path_str: String = path?.to_string_lossy().to_string();
+
+            let Some(cap) = re.captures(&path_str) else {
+                continue;
+            };
+
+            // Insert into the cache
+            name_ver_tree.insert(
+                format!("{}-{}", &cap["name"].to_string(), &cap["ver"].to_string()),
+                cap["hash"].into(),
+            );
```
This still seems like a cache implementation builtin to the logic since it applies a copy.

Synicix replied:
The comment is stale, it still uses the same structure as before, just without storing it in memory. The Btreemap is just a nice way to quickly look up annotation-version -> hash. I will update the comment

guzman-raphael commented on src/store/filestore.rs:
```diff
@@ -1,213 +1,216 @@
-    fn save_pod(&self, pod: &Pod) -> Result<(), Box<dyn Error>> {
-        let class = "pod";
+    fn save_pod(&self, pod: &Pod) -> Result<()> {
+        self.save_item(pod, &pod.hash, pod.annotation.as_ref())
```
`item` might not be clear/intuitive. We've been calling the core classes *models* which was borrowed from [Pydantic](https://docs.pydantic.dev/latest/concepts/models/) community. Perhaps we can rename `item` to `instance` similarly to how they refer to it?

Synicix replied:
Not sure instance is a good name for this either since isn' that also pretty generic too?

eywalker replied:
we could just use the term `model` no?

guzman-raphael replied:
Discussed with @Synicix. Planning to go with `item` -> `model` and `ItemKey` -> `ModelID`.

eywalker replied:
Sounds good!

guzman-raphael commented on src/store/filestore.rs:
```diff
@@ -1,213 +1,216 @@
+    ) -> Result<()> {
+        // Save the item first
+        Self::save_file(
+            &self.make_path::<T>(hash, "spec.yaml"),
```
The intent behind `make_annotation_path`, `make_spec_path`, and `parse_annotation_path` was so that there was no other mention of paths anywhere so that any change to paths or to file naming would just affect those methods so it takes minimal work to make updates later.

It seems odd that we expect the caller to *know* the file is called `spec.yaml` which also breaks in more places if we change this later.

Synicix replied:
Updated with static const now

guzman-raphael commented on src/store/filestore.rs:
```diff
@@ -1,213 +1,216 @@
-            false,
-        )?;
+    /// Return the name version index btree where key is name-version and value is hash of pod
+    fn list_pod(&self) -> Result<BTreeMap<String, String>> {
```
Similar to my previous feedback [here](https://github.com/walkerlab/orcapod/pull/34#discussion_r1812862407), I still prefer having tabular data represented as a `BTreeMap<String, Vec<_>>` where the key is column name and the vector is the values for the rows.

Synicix replied:
One problem with this is the vectors will be locked a specific type such as String. I assume we are okay with that? A vector of struct would be more flexiable in the the future instead of a btreemap

Synicix replied:
Chhange to vector of struct

guzman-raphael commented on src/store/filestore.rs:
```diff
@@ -1,213 +1,216 @@
+    pub fn new(directory: impl AsRef<Path>) -> Self {
         Self {
-            directory: directory.into(),
+            directory: directory.as_ref().into(),
```
Why is this better? Is there some documentation you could point me to?

Synicix replied:
We don't really want to force it to covert into PathBuf, instead we just want it to implment the Path reference.
Pathbuf is like String and path is like &str. We got the idea from how std::path lib implments it: https://doc.rust-lang.org/src/std/path.rs.html#3262

eywalker replied:
Yeah this allows for the method to more broadly accept different input types. It is a bit more idiomatic to use the following: PathBuf::from(directory.as_ref())

guzman-raphael commented on src/store/filestore.rs:
```diff
@@ -1,213 +1,216 @@
+
+        let search_pattern = self.make_dir_path::<T>("*").join("annotations/*");
+
+        for path in glob::glob(&search_pattern.to_string_lossy())? {
```
In general, it is *usually* better to perform an iterator (lazy) approach as opposed to explicit loops. Here are a few reference articles for Rust specifically that can give some context where I am coming from.

- https://www.lurklurk.org/effective-rust/iterators.html
- https://doc.rust-lang.org/book/ch13-04-performance.html

Synicix replied:
I prefer readability in this case since I didn't want to redefine the regex pattern every loop like you had in your previous example. Plus it had that odd return type.

eywalker replied:
Let's put this as a point for a quick discussion

guzman-raphael replied:
@guzman-raphael Will look into a few cases and report back.

eywalker replied:
In the interest of having some closure on this PR that's been up for a while, let's set this as a separate issue and address in the future.

guzman-raphael reviewed (changes requested):
Updating my review. Haven't gone through the tests much since I'll probably do a follow-up pass over it after merge anyway to help out.

Synicix committed:
Merge remote-tracking branch 'upstream/dev' into parse_annotation

Synicix committed:
Merge branch 'raphael' into parse_annotation

Synicix committed:
Merge remote-tracking branch 'upstream/dev' into parse_annotation

Synicix committed:
Switch to new error system

Synicix committed:
Change to new error handling and fix certain recommendations

Synicix commented:
@eywalker @guzman-raphael Ready for review again, but probably need to discuss on some of the points above.

eywalker commented on src/error.rs:
```diff
@@ -8,10 +9,9 @@ use std::{
 };
 /// Shorthand for a Result that returns an `OrcaError`.
-pub(crate) type Result<T> = result::Result<T, OrcaError>;
+pub type Result<T> = anyhow::Result<T, OrcaError>;
```
Let's discuss the use of `anyhow` crate along with whether to make this alias public.

eywalker replied:
I'm going to let this pass for now and we can decide to revisit the use of `anyhow` in the future PR. Fortunately, the usage pattern of `Result<T>` won't change.

eywalker commented on src/error.rs:
```diff
@@ -60,12 +60,17 @@ impl Display for OrcaError {
             Kind::NoAnnotationFound(class, name, version) => {
                 write!(f, "No annotation found for `{name}:{version}` {class}.")
             }
-            Kind::NoRegexMatch => {
```
Why was this removed?

Synicix replied:
Does regex function I rewrote allow for no mismatch and relies on it. So instead of erroring out it just returns and empty Btree

guzman-raphael commented on src/store/filestore.rs:
```diff
@@ -1,207 +1,256 @@
+        clippy::unwrap_used,
+        reason = "It shouldn't failed as we added the keys ourselves at the start"
+    )]
+    fn list_pod(&self) -> Result<BTreeMap<String, Vec<String>>> {
```
@Synicix Is going to change the return type to a `Vec<Struct>` and we'll revisit once the update is in.

Synicix replied:
Done

Synicix committed:
Fix a few things based on Rapahel feedback

Synicix committed:
Remove accidently commited files

Synicix committed:
Merge branch 'parse_annotation' into pod-job

Synicix committed:
Add pod job and tests

Synicix committed:
Clean up a few stuff for incoming pod_job

Synicix commented:
@guzman-raphael @eywalker So with the discussion that Rapahel and me had. I resolve some of the change requested with a few remaining.

- Unable to get the anyhow alias to work due to the fact that the return error type must be of anyhow::Error for the stack trace to work. Thus it has to be used on the top level for it to display to force conversion of our OrcaError into anyhow::Error
- Waiting for Raphael to see if there is a better implemnetation with maps regarding the name_ver_tree
- Test still doesn't do the fixture for the pod yet (might be able to pick this up later tonight or tomrrow)

eywalker commented on src/store/mod.rs:
```diff
@@ -1,5 +1,22 @@
-use std::collections::BTreeMap;
+
+/// Enum for identification to
+pub enum ModelID {
```
We should consider making this clonable

eywalker commented on src/store/mod.rs:
```diff
@@ -15,20 +32,26 @@ pub trait Store {
     ///
     /// Will return `Err` if there is an issue querying metadata from existing pods in the store.
-    fn list_pod(&self) -> Result<BTreeMap<String, Vec<String>>>;
+    fn list_pod(&self) -> Result<Vec<ModelInfo>>;
```
Let's follow up on the specification of list behavior as a dedicated issue.

eywalker reviewed (approved):
There are a few hanging points that I'd want to move as issues to be followed up on so that this PR can be merged. This branch has already been serving as the base for many development and it should be merged.

guzman-raphael reviewed (approved):
Merging for now to unblock and address the remaining points in a follow-up PR.
